server/embedded/src/org/labkey/embedded/LabKeyServer.java

```diff
@@ -0,0 +1,358 @@
+				}
+
+				// TODO : 8021 :fix context path - put app at root by default
+				StandardContext context = (StandardContext)tomcat.addWebapp("/labkey", webAppLocation);
+
+                // Push the JDBC connection for the primary DB into the context so that the LabKey webapp finds them
+				getDataSourceResources(props).forEach(contextResource -> context.getNamingResources().addResource(contextResource));
+				// Add the SMTP config
+				context.getNamingResources().addResource(getMailResource());
+
+				// And the master encryption key
+				context.addParameter("MasterEncryptionKey", props.getMasterEncryptionKey());
+
+				// Point at the special classloader with the hack for SLF4J
+				WebappLoader loader = new WebappLoader();
+				loader.setLoaderClass(LabKeySpringBootClassLoader.class.getName());
+				context.setLoader(loader);
+				context.setParentClassLoader(this.getClass().getClassLoader());
+
+				return super.getTomcatWebServer(tomcat);
+			}
+
+			private List<ContextResource> getDataSourceResources(ContextProperties props)
+			{
+				List<ContextResource> dataSourceResources = new ArrayList<>();
+				var numOfDataResources = props.getUrl().size();
+
+				for (int i = 0; i < numOfDataResources; i++)
+				{
+					ContextResource dataSourceResource = new ContextResource();
+					dataSourceResource.setName(props.getDataSourceName().get(i));
+					dataSourceResource.setAuth("Container");
+					dataSourceResource.setType(DataSource.class.getName());
+					dataSourceResource.setProperty("driverClassName", props.getDriverClassName().get(i));
+					dataSourceResource.setProperty("url", props.getUrl().get(i));
+					dataSourceResource.setProperty("password", props.getPassword().get(i));
+					dataSourceResource.setProperty("username", props.getUsername().get(i));
+					dataSourceResource.setProperty("maxTotal", "20");
+					dataSourceResource.setProperty("maxIdle", "10");
+					dataSourceResource.setProperty("maxWaitMillis", "120000");
+					dataSourceResource.setProperty("accessToUnderlyingConnectionAllowed", "true");
+					dataSourceResource.setProperty("validationQuery", "SELECT 1");
+					dataSourceResources.add(dataSourceResource);
+				}
+
+				return  dataSourceResources;
+			}
+
+			private ContextResource getMailResource()
+			{
+				// Get session/mail properties
+				MailProperties mailProps = smtpSource();
+				ContextResource mailResource = new ContextResource();
+				mailResource.setName("mail/Session");
+				mailResource.setAuth("Container");
+				mailResource.setType("javax.mail.Session");
+				mailResource.setProperty("mail.smtp.host", mailProps.smtpHost);
+				mailResource.setProperty("mail.smtp.user", mailProps.smtpUser);
+				mailResource.setProperty("mail.smtp.port", mailProps.smtpPort);
+				return mailResource;
+			}
+		};
+	}
+
+	/**
+	 * Extracts a zip file specified by the zipFilePath to a directory specified by
+	 * destDirectory (will be created if does not exists)
+	 * @param zipInputStream
```

Comment:
Useless JavaDoc - delete these three lines or add info about the params and exception